src/signing/env_mnemonic.rs

```diff
@@ -0,0 +1,161 @@
+    pbkdf2::pbkdf2::<Hmac<sha2::Sha512>>(input, salt.as_bytes(), PBKDF2_ROUNDS, &mut seed);
+    seed
+}
+
+fn mnemonic_to_ed25_seed(mnemonic: String, password: String) -> ed25519::Ed25519Seed {
+    let salt = format!("mnemonic{}", password);
+    let normalized_salt = salt.nfkd().to_string();
+    let bytes = _pbkdf2(mnemonic.as_bytes(), &normalized_salt);
+    ed25519::Ed25519Seed::from_bytes(&bytes).unwrap()
+}
+
+fn derive_into_address(private_key: ed25519::Ed25519PrivateKey) -> String {
+    let public_key = private_key.generate_public_key().to_bytes();
+    // Hash the public key to get the address
+    let mut hasher = VarBlake2b::new(32).unwrap();
+    hasher.update(public_key);
+    let mut result = vec![];
+    hasher.finalize_variable(|res| {
+        result = res.to_vec();
+    });
+
+    let prefix: Vec<u8> = vec![1]; // prefix for ed25 addresses
+    let mut data = prefix;
+    data.append(&mut result);
+    bech32::encode("iota", data.to_base32()).unwrap()
+}
+
+#[derive(Default)]
+pub struct EnvMnemonicSigner;
+
+impl EnvMnemonicSigner {
+    fn get_seed(&self) -> ed25519::Ed25519Seed {
+        mnemonic_to_ed25_seed(
+            env::var("IOTA_WALLET_MNEMONIC")
+                .expect("must set the IOTA_WALLET_MNEMONIC environment variable"),
+            env::var("IOTA_WALLET_MNEMONIC_PASSWORD").unwrap_or_else(|_| "password".to_string()),
+        )
+    }
+
+    fn get_private_key(
+        &self,
+        derivation_path: String,
+    ) -> crate::Result<ed25519::Ed25519PrivateKey> {
+        let seed = self.get_seed();
+        Ok(ed25519::Ed25519PrivateKey::generate_from_seed(
+            &seed,
+            &BIP32Path::from_str(&derivation_path).map_err(|e| anyhow::anyhow!(e.to_string()))?,
+        )
+        .map_err(|e| anyhow::anyhow!(e.to_string()))?)
+    }
+}
+
+impl super::Signer for EnvMnemonicSigner {
+    fn init_account(&self, account: &Account, mnemonic: Option<String>) -> crate::Result<String> {
+        if let Some(mnemonic) = mnemonic {
+            env::set_var("IOTA_WALLET_MNEMONIC", mnemonic);
+        }
+        Ok(thread_rng().gen_ascii_chars().take(10).collect())
+    }
+
+    fn generate_address(
+        &self,
+        account: &Account,
+        address_index: usize,
+        internal: bool,
+    ) -> crate::Result<iota::Address> {
+        let private_key = self.get_private_key(format!(
+            "m/44H/4218H/{}H/{}H/{}H",
+            account.index(),
+            internal as u32,
+            address_index
+        ))?;
+        let address_str = derive_into_address(private_key);
+        crate::address::parse(address_str)
+    }
+
+    fn sign_message(
+        &self,
+        account: &Account,
+        essence: &iota::TransactionEssence,
+        inputs: &mut Vec<super::TransactionInput>,
+    ) -> crate::Result<Vec<iota::UnlockBlock>> {
+        let mut serialized_essence = Vec::new();
+        essence
+            .pack(&mut serialized_essence)
```

Comment:
You can use `pack_new` instead, it's more efficient as it does less memory allocations